Rename all the service reference things

dougbu opened the pull request:
- #7492
- remove document generation from client project; will be included in Web API project infrastructure (coming soon)

Also
- provide `%(FirstForGenerator)` metadata, #4916
- add `$(OpenApiGenerateAtDesignTime)` property (default `true` for now), #4944
- generate code in `obj` directory by default, #4945

dougbu committed:
Rename all the service reference things
- #7492
- remove document generation from client project; will be included in Web API project infrastructure (coming soon)

Also
- provide `%(FirstForGenerator)` metadata, #4916
- add `$(OpenApiGenerateAtDesignTime)` property (default `true` for now), #4944
- generate code in `obj` directory by default, #4945

dougbu commented:
Pity GitHub doesn't recognize the .props and .targets files were renamed before things in them were renamed. If you're primarily interested in the delta from existing code, I suggest
``` PowerShell
git checkout dougbu/code.generation.renames.7492
git show --find-renames=20 -- *.props
git show --find-renames=20 -- *.targets
```

dougbu commented:
@natemcmaster I know you found one typo. Anything else ?

dougbu committed:
Fix typo, update solution

natemcmaster commented on src/Mvc/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.targets:
```diff
@@ -0,0 +1,151 @@
+      AfterTargets="ResolveProjectReferences"
+      Condition="'%(ProjectReferenceWithConfiguration.OpenApiReference)' == 'true' AND '$(DesignTimeBuild)' != 'true' AND '$(BuildingProject)' == 'true'"
+      Inputs="%(ProjectReferenceWithConfiguration.Identity)"
+      Outputs="&lt;not-a-file !&gt;">
```
This is a unsupported usage for inputs and outputs. It looks to me like you're trying to batch over ProjectReferenceWithConfiguration. You don't have to do this with target batching.

natemcmaster commented on src/Mvc/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.targets:
```diff
@@ -0,0 +1,151 @@
+    </PropertyGroup>
+
+    <MSBuild Target="OpenApiGetDocuments"
+        Project="$(_Identity)"
```
```suggestion
        Project="@(ProjectReferenceWithConfiguration)"
```

No need to create an intermediate property.

natemcmaster commented on src/Mvc/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.targets:
```diff
@@ -0,0 +1,151 @@
+        Properties="%(ProjectReferenceWithConfiguration.SetConfiguration); %(ProjectReferenceWithConfiguration.SetPlatform); %(ProjectReferenceWithConfiguration.SetTargetFramework)"
+        RebaseOutputs="true"
+        RemoveProperties="%(ProjectReferenceWithConfiguration.GlobalPropertiesToRemove);TargetFrameworks;RuntimeIdentifier">
+      <Output TaskParameter="TargetOutputs" PropertyName="_Temporary" />
```
There are some MSBuild way of doing this which don't require manually matching up some returned metadata from  a target. I would start by changing this to return an item instead of a property.

```suggestion
      <Output TaskParameter="TargetOutputs" ItemName="_ResolvedOpenApiProjectWithDocuments" />
```

Next, make the target you're invoking, in this case "OpenApiGetDocuments", look like this:

```xml
<Target Name="OpenApiGetDocuments"
              Returns="@(OpenApiProjectWithDocuments)">
   <ItemGroup>
       <OpenApiProjectWithDocuments Include="$(MSBuildProjectFullPath)">
          <Documents>$(stuffHere)</Documents>
      </OpenApiProjectWithDocuments>
  </ItemGroup>
</Target>
```

natemcmaster commented on src/Mvc/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.targets:
```diff
@@ -0,0 +1,151 @@
+    </MSBuild>
+
+    <ItemGroup>
+      <OpenApiProjectReference Update="$(_Identity)">
```
This doesn't work during target execution. It's one of those quirks. `Update` is only allowed during static evaluation. I believe newer versions of MSBuild will properly throw an error here. Older versions allowed this, but behave incorrectly.

natemcmaster commented on src/Mvc/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.targets:
```diff
@@ -0,0 +1,151 @@
+        Extension="$(DefaultLanguageSourceExtension)"
+        Namespace="$(RootNamespace)"
+        OutputDirectory="$(OpenApiDefaultOutputDirectory)">
+      <Output TaskParameter="Outputs" ItemName="_Temporary" />
```
```suggestion
      <Output TaskParameter="Outputs" ItemName="OpenApiReferenceWithMetadata" />
```

A good practice in MSBuild is to use new item groups as you transform the data, instead of updating the items in place. This helps with extensibility because then people extending your targets don't need to worry so much about the order in which targets execute.

natemcmaster commented on src/Mvc/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.targets:
```diff
@@ -0,0 +1,151 @@
+          Condition="Exists('%(OpenApiReference.OutputPath)') AND ! $([System.IO.File]::Exists('%(OpenApiReference.OutputPath)'))" />
+
+      <!-- If OutputPath is a file, add it directly to relevant items. -->
+      <TypeScriptCompile Remove="@(_Files)" Condition="'%(_Files.OutputPathExtension)' == '.ts'" />
```
Don't forget about `.tsx` :)

natemcmaster commented on src/Mvc/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.targets:
```diff
@@ -0,0 +1,151 @@
+      </Compile>
+
+      <!-- Otherwise, add all descendant files with the expected extension. -->
+      <TypeScriptCompile Remove="@(_Directories -> '%(Identity)/**/*.ts')" />
```
I'm assuming you're doing the `Remove` to avoid duplicate entries, right? A more efficient way to do this is in MSBuild is
```xml
<TypeScriptCompile Include=" new pattern here " Exclude="@(TypeScriptCompile)" />
```

natemcmaster reviewed (commented):
Only had time to look at Microsoft.Extensions.ApiDescription.Client.targets.

dougbu commented:
@glennc this PR partially addresses #7491 because it provides a default `%(CodeGenerator)` value. Does not add proposed `%(Language)` metadata.

dougbu committed:
Fix build and address PR comments
- adjust eng/ProjectReferences.props to new project name
- simplify `_CreateOpenApiReferenceItemsForOpenApiProjectReferences` target e.g. remove batching
- simplify item de-duplication in `_CreateCompileItemsForServiceFileReferences` target i.e. use `Remove` attribute
  - also handle `.tsx` files in this target
- remove malformed `Exists(...)` check in `_GenerateOpenApiReferenceCode` target

nits:
- remove a useless `StringBuilder.Append(...)` call
- remove `%(OpenApiProjectReference.SourceProject)` metadata, duplicated `%(OriginalItemSpec)`
- be more consistent about using element syntax for item metadata

dougbu committed:
Fixup included #4916 fix
- set `%(FirstForGenerator)` based on `%(CodeGenerator)` metadata

Side note, PR fixes include
- add `%(FirstForGenerator)` metadata, #4916
- add support for disabling design-time builds, #4944
- change default output directory to `obj/`, #4945 1 of 2 (remainder will come with #8242)
- provide a default `%(CodeGenerator)` value, ##7491 1 of 2 (remainder will come in next milestone)

pranavkm commented on src/Mvc/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.targets:
```diff
@@ -0,0 +1,149 @@
+
+      <Compile Include="@(_Files)"
+          Exclude="@(Compile)"
+          Condition="'$(DefaultLanguageSourceExtension)' != '.ts' AND '%(_Files.OutputPathExtension)' == '$(DefaultLanguageSourceExtension)'">
```
Have we tested this with F# projects? If not, should we limit this to 'Condition="'$(Language)'=='C#'"`?

dougbu replied:
This condition is correct because compiled languages other than TypeScript are not mixed into other projects. That is, this is about the _exception_ to normal mono-language projects.

pranavkm commented on src/Mvc/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.targets:
```diff
@@ -0,0 +1,149 @@
+      Do not change %(ReferenceOutputAssembly) if project contains duplicate @(ProjectReference) and
+      @(OpenApiProjectReference) items.
+    -->
+    <ProjectReference Include="@(OpenApiProjectReference)" Exclude="@(ProjectReference)">
```
Does this work if the project have different TFMs? Like does it work if you reference a `netcoreapp` project from a `netstandard` project? I know we struggled with this when we were authoring test projects in Razor.

dougbu replied:
It doesn't work perfectly yet and will be refined. As-is, it brings a large reliability improvement over explicit `<MSBuild />` tasks to restore and build the server project.

For later consideration, could you point me to the Razor solution to this issue?

pranavkm commented on src/Mvc/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.targets:
```diff
@@ -0,0 +1,149 @@
+      <NoWarn>NU1702</NoWarn>
+      <ReferenceOutputAssembly>false</ReferenceOutputAssembly>
+    </ProjectReference>
+    <ProjectReference Update="@(OpenApiProjectReference)">
```
Any reason not to inline this with the previous include?

dougbu replied:
Because when `@(ProjectReference)` and `@(OpenApiProjectReference)` overlap we need to get the documents. The previous element is about the non-overlapping case and this covers everything.

natemcmaster replied:
MSBuild and the SDK don't handle duplicate `<ProjectReference>` items well. This can cause some strange behaviors such as https://github.com/Microsoft/msbuild/issues/2688. Consider making this overlap an error.

dougbu replied:
The two blocks are separate specifically to avoid creating duplicate `<ProjectReference>` items. Why make overlap between `@(ProjectReference)` and `@(OpenApiProjectReference)` an error with that in place?

pranavkm commented on src/Mvc/Extensions.ApiDescription.Client/src/Microsoft.Extensions.ApiDescription.Client.nuspec:
```diff
@@ -21,9 +21,5 @@
     <file src="buildMultiTargeting\*" target="buildMultiTargeting" />
     <file src="bin\$configuration$\net461\Microsoft.Extensions.ApiDescription.Tasks.*" target="tasks\net461" />
     <file src="bin\$configuration$\netstandard2.0\Microsoft.Extensions.ApiDescription.Tasks.*" target="tasks\netstandard2.0" />
-    <file src="..\..\dotnet-getdocument\src\bin\$configuration$\netcoreapp2.1\publish\*.*" target="tools" />
```
What happened to these?

dougbu replied:
They aren't necessary with the new client / server split. The tool will come back in Microsoft.Extensions.ApiDescription.Server

pranavkm commented on src/Mvc/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.targets:
```diff
@@ -0,0 +1,149 @@
+    </GetFileReferenceMetadata>
+
+    <ItemGroup>
+      <OpenApiReference Remove="@(OpenApiReference)" />
```
Should this be a different itemgroup rather than re-using `OpenApiReference`? It would be easier to debug this and wouldn't modify a user specified item.

dougbu replied:
No, this is intended to flow `@(OpenApiProjectReference)` items into `@(OpenApiReference)`. Besides, modifying user-specified items e.g. `@(Compile)` is just normal.

natemcmaster commented on src/Mvc/Extensions.ApiDescription.Client/src/build/Microsoft.Extensions.ApiDescription.Client.props:
```diff
@@ -0,0 +1,67 @@
+      absolute path.
+    -->
+    <OpenApiDefaultOutputDirectory
+        Condition="'$(OpenApiDefaultOutputDirectory)' == ''">$(BaseIntermediateOutputPath)</OpenApiDefaultOutputDirectory>
```
FYI - BaseIntermediateOutputPath might not be fully set when this file is evaluated. Would recommend moving this to a .targets file instead. BaseIntermediateOutputPath isn't defined until Microsoft.NET.DefaultOutputPaths.targets is evaluated, which happens at the end of project evaluation. Props file evaluation happens near the beginning of project eval.

dougbu replied:
Hmm, it's working at the moment. I'll move it if we see issues or if I otherwise _must_ update this PR.

natemcmaster reviewed (commented):
A few suggestions:
